aten/src/ATen/native/cuda/ReduceSumProdKernel.cu

```diff
@@ -61,7 +69,28 @@ static void prod_kernel_cuda(TensorIterator& iter) {
   });
 }
 
+static void nanprod_kernel_cuda(TensorIterator& iter) {
```

Comment:
This code looks very similar to `prod_kernel_cuda`, which makes sense, but is there a way to avoid the code redundancy? Maybe these functions could share a templated helper function, for example?

Reply:
Done, thanks to @kshitij12345 for his implementation in #38628. The changes are highly motivated from his work in `nansum` PR.